utils/expire_map.go

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package utils
+
+import (
+	"sync"
+	"time"
+)
+
+const NeverExpire time.Duration = -1
+
+type expiredata struct {
+	data        interface{}
+	expiredTime int64
+	valid       int64
+}
+
+func (d *expiredata) checkValid() bool {
+	// if valid is equal NeverExpire and don't need to update.
+	if d.valid == int64(NeverExpire) {
+		return true
+	}
+
+	return d.expiredTime > time.Now().Unix()
+}
+
+type ExpiredMap struct {
+	syncMap *sync.Map
+	syncMod bool // true means using synchronous update mode, otherwise async mod
+
+	// When the cache expires, it is used to update the cache.
+	UpdateHandler func(interface{}) (interface{}, bool)
+}
+
+// handler is used to update the data if the cache is invalid during Get.
+// syncMod is set true means that the handler is called synchronously, and the others are asynchronous.
+func NewExpiredMap(handler func(interface{}) (interface{}, bool), syncMod bool) *ExpiredMap {
+	return &ExpiredMap{
+		syncMap:       &sync.Map{},
+		UpdateHandler: handler,
+		syncMod:       syncMod,
+	}
+}
+
+// key and val represent cached index and user data.
+// valid is used to set the expire time of the cache. For example, if valid=10 means the data expires after 10s.
```

Comment:
这个注释好像不太对， time.Duration的话，是一个int64， 10 就是10ns，这里应该是10 * time.Second.我看这里要转换成int64，如果单位是s的话，应该是不对的？

Reply:
最终内部封装使用的地方其实还是使用的是秒（s）
```
func (t Time) Unix() int64
Unix returns t as a Unix time, the number of seconds elapsed since January 1, 1970 UTC. The result does not depend on the location associated with t.
```